### PR TITLE
Heartbeat v2 monitor better filtering

### PR DESCRIPTION
**Pull request by iulianpascalau**
## Description of the reasoning behind the pull request (what feature was missing / how the problem was manifesting itself / what was the motive behind the refactoring)
- after some **mainnet** node upgrades, the observed heartbeat status response randomly showed active or inactive nodes in the network. This was caused by the fact that the node returned all messages (inactive or active) regardless that the public key is now behind another pid
  
## Proposed Changes
- made the heartbeat v2 monitor struct better filter-out inactive messages by using the following algorithm:

suppose that, for the same BLS public key we got `n` messages from `n` different pids:
1. if all stored messages show inactive nodes, only keep the latest message, remove the older ones, and put the value `0` on the `NumInstances` field from the `data.PubKeyHeartbeat` struct;
2. if all stored messages show only active nodes, sort the messages based on the `pid` (so that the response will be consistent on all requests), take the first active heartbeat message, and set the `NumInstances` field from `data.PubKeyHeartbeat` struct to how many active messages the node has;
3. if some stored messages show inactive nodes and others show active nodes, remove the stored messages that show inactive nodes, and for the rest of the messages that show active nodes, apply point 2. 

## Testing procedure
- standard system test
- mainnet-specific tests that will prove that the new code will better filter-out inactive messages

### CODE REVIEW

File: heartbeat/monitor/heartbeatMessage.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+		activePids:         make([]core.PeerID, 0),
+		activeHeartbeats:   make(map[core.PeerID]*data.PubKeyHeartbeat),
+		inactiveHeartbeats: make(map[core.PeerID]*data.PubKeyHeartbeat),
+	}
+}
+
+func (instance *heartbeatMessages) addMessage(pid core.PeerID, message *data.PubKeyHeartbeat) {
+	if message.IsActive {
+		instance.activePids = append(instance.activePids, pid)
+		instance.activeHeartbeats[pid] = message
```
Comment (dragos-rebegea) [inline heartbeat/monitor/heartbeatMessage.go:27]:
```suggestion
                if (pid < instance.activePid) {
		    instance.activePid = pid
		    instance.activeHeartbeat = message
		}
```
this way you just store the message for the lowerst peerID, so you dont have to sort to find it after that

**Review by sstanculeanu — COMMENTED** [review]
no other comments

File: heartbeat/monitor/heartbeatMessage.go (hunk 1 above)
Comment (iulianpascalau) [reply heartbeat/monitor/heartbeatMessage.go:27]: refactored, now the `heartbeatMessages` struct does not contain any maps. Removed several for... loops in the process. 
Good suggestion :+1:

**Review by dragos-rebegea — APPROVED** [review]

File: heartbeat/monitor/heartbeatMessage.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
+package monitor
+
+import (
+	"github.com/ElrondNetwork/elrond-go-core/core"
+	"github.com/ElrondNetwork/elrond-go/heartbeat/data"
+)
+
+type heartbeatMessages struct {
+	firstActivePid          core.PeerID
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply heartbeat/monitor/heartbeatMessage.go:9]: yes, but we need a decode for that as it is already base58 encoded. I would keep this as it is for simplicity and ease of understanding

**Review by sstanculeanu — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed.

**Pull request merged**